# [5주차 기본과제] 뉴스 기사 분류를 Gemma로 구현하기


In [88]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("google/gemma-2-2b-it")
model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2-2b-it",
    device_map="auto",
    torch_dtype=torch.bfloat16,
)

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: fineGrained).
Your token has been saved to /Users/obov/.cache/huggingface/token
Login successful


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/636 [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [89]:
from datasets import load_dataset


news_ds = load_dataset("fancyzhx/ag_news")
news_ds

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 7600
    })
})

In [90]:
world_ds = news_ds["train"].filter(lambda data: data["label"] == 0)
world_ds["text"][:3]

['Venezuelans Vote Early in Referendum on Chavez Rule (Reuters) Reuters - Venezuelans turned out early\\and in large numbers on Sunday to vote in a historic referendum\\that will either remove left-wing President Hugo Chavez from\\office or give him a new mandate to govern for the next two\\years.',
 'S.Koreans Clash with Police on Iraq Troop Dispatch (Reuters) Reuters - South Korean police used water cannon in\\central Seoul Sunday to disperse at least 7,000 protesters\\urging the government to reverse a controversial decision to\\send more troops to Iraq.',
 "Palestinians in Israeli Jails Start Hunger Strike (Reuters) Reuters - Thousands of Palestinian\\prisoners in Israeli jails began a hunger strike for better\\conditions Sunday, but Israel's security minister said he\\didn't care if they starved to death."]

In [91]:
sports_ds = news_ds["train"].filter(lambda data: data["label"] == 1)
sports_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [92]:
business_ds = news_ds["train"].filter(lambda data: data["label"] == 2)
business_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [93]:
sci_tech_ds = news_ds["train"].filter(lambda data: data["label"] == 3)
sci_tech_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [94]:
news_ds["train"].features

{'text': Value(dtype='string', id=None),
 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}

In [95]:
label_map = news_ds["train"].features["label"].names

In [97]:
news_ds["train"].info

DatasetInfo(description='', citation='', homepage='', license='', features={'text': Value(dtype='string', id=None), 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}, post_processed=None, supervised_keys=None, builder_name='parquet', dataset_name='ag_news', config_name='default', version=0.0.0, splits={'train': SplitInfo(name='train', num_bytes=29832303, num_examples=120000, shard_lengths=None, dataset_name='ag_news'), 'test': SplitInfo(name='test', num_bytes=1880424, num_examples=7600, shard_lengths=None, dataset_name='ag_news')}, download_checksums={'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/train-00000-of-00001.parquet': {'num_bytes': 18585438, 'checksum': None}, 'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/test-00000-of-00001.parquet': {'num_bytes': 1234829, 'checksum': None}}, download_size=19820267, post_processing_size=None, dataset_size=31712727, size_in_bytes=51532994)

In [98]:
from pprint import pprint

pprint(news_ds["train"].info)

DatasetInfo(description='',
            citation='',
            homepage='',
            license='',
            features={'label': ClassLabel(names=['World',
                                                 'Sports',
                                                 'Business',
                                                 'Sci/Tech'],
                                          id=None),
                      'text': Value(dtype='string', id=None)},
            post_processed=None,
            supervised_keys=None,
            builder_name='parquet',
            dataset_name='ag_news',
            config_name='default',
            version=0.0.0,
            splits={'test': SplitInfo(name='test',
                                      num_bytes=1880424,
                                      num_examples=7600,
                                      shard_lengths=None,
                                      dataset_name='ag_news'),
                    'train': SplitInfo(name='train',
   

In [99]:
tokenizer("World", return_tensors="pt")
tokenizer("World", return_tensors="pt")["input_ids"].shape[-1]

2

In [100]:
label_sentences = [
    "Batman: This is mainly about the world, the one we’re sworn to protect.",
    "Joker: Why so serious? Clearly related to Sports news!",
    "Penguin: It’s all about Business, and trust me, I know how to play that game.",
    "Bane: This article mainly covers Sci/Tech.",
]

task_description = """
Task Description:
The scene unfolds with four iconic characters from the Batman universe seated around a table, each reading the same article. Batman, ever composed and logical, is the first to speak, delivering a clear and concise assessment of the piece. The Joker follows, making a mocking and unpredictable remark, turning the topic on its head. Two-Face, as always, relies on the flip of a coin to determine his take. Finally, Bane, with his focus on power and control, offers his perspective, emphasizing the article’s deeper implications.

Your task is simple: “Who explained the article the best?”.

Now, it’s time to choose—who described it best?

ARTICLE:"""

In [101]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.mps.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

In [102]:
news_ds["train"][1]["text"]

'Carlyle Looks Toward Commercial Aerospace (Reuters) Reuters - Private investment firm Carlyle Group,\\which has a reputation for making well-timed and occasionally\\controversial plays in the defense industry, has quietly placed\\its bets on another part of the market.'

In [103]:
news_ds["train"][1]["label"]

2

In [104]:
text_label = news_ds["train"][179]

probs = zero_shot_classification(
    text_label["text"],
    task_description,
    label_sentences,
)
print(probs)
import numpy as np

print(text_label["label"])
np.argmax(np.array(probs)).item()

[99.28125, 70.71484375, 158.25, 48.431640625]
3


2

In [86]:
def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_news_ds = news_ds.map(preprocess_function, batched=True)
tokenized_news_ds

Map:   0%|          | 0/120000 [00:00<?, ? examples/s]

Map:   0%|          | 0/7600 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 7600
    })
})

In [105]:
import numpy as np
from tqdm import tqdm
from colorama import Fore, Style, Back
import torch.nn.functional as F
import humanize
import math
import random


def calculate_nll(probs_softmax, label):
    return -torch.log(probs_softmax[label])


def get_confidence(probs_softmax):
    top_two_probs, _ = torch.topk(probs_softmax, 2)
    # 가장 높은 확률을 두 번째로 높은 확률로 나누어 비율을 계산
    return math.log10(top_two_probs[0] / top_two_probs[1])


total = 50
n_corrects = 0

# 0부터 7600 사이에서 겹치지 않게 50개의 숫자 뽑기
random_numbers = random.sample(range(7601), total)

for i in tqdm(range(total)):

    print(f"{Fore.YELLOW}----------------------------------{Fore.RESET}")
    text = tokenized_news_ds["test"][random_numbers[i]]["text"]
    label = tokenized_news_ds["test"][random_numbers[i]]["label"]
    probs = zero_shot_classification(
        text,
        task_description,
        label_sentences,
    )

    probs_tensor = torch.tensor(probs)
    probs_softmax = F.softmax(probs_tensor, dim=0)

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1
        tqdm.write(f"Index {i}: {Fore.GREEN}Success 🎉 {Style.RESET_ALL} ")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
    else:
        tqdm.write(f"Index {i}: {Fore.RED}Fail 🧐 {Style.RESET_ALL}")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
        print(f"Failed Query: {Back.BLACK}{text}{Back.RESET}")
        label, pred = label_map[label], label_map[pred]
        print(
            f"Label : {Fore.CYAN} {label} {Fore.RESET} Pred : {Fore.MAGENTA} {pred} {Fore.RESET}"
        )
    current_accuracy = n_corrects / (i + 1)
    print(f"Current Accuracy: {Fore.BLUE}{current_accuracy:,.2%}{Fore.RESET}")

print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")
print(
    f"Total Correct Predictions: {Style.BRIGHT}{n_corrects} ({n_corrects/total:.2%}){Style.RESET_ALL}"
)
print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")

  0%|          | 0/50 [00:00<?, ?it/s]

----------------------------------


  2%|▏         | 1/50 [00:04<03:27,  4.24s/it]

Index 0: Fail 🧐 
dataset index: 5093
Confidence: 26.28
NLL: inf
Failed Query: Hubble Sees Rare Triple Jupiter Eclipse Nov. 4, 2004 - A rare alignment of Jupiter #39;s three largest moons across the planet #39;s face was captured on film by the Hubble Space Telescope.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


  4%|▍         | 2/50 [00:07<02:44,  3.43s/it]

Index 1: Fail 🧐 
dataset index: 5425
Confidence: 25.25
NLL: 84.63
Failed Query: Mayor sure of ballpark support DC Mayor Anthony A. Williams said yesterday he is  quot;very confident quot; that he has the seven necessary votes from the DC Council for his plan to build a ballpark near South Capitol Street Southeast.
Label :  Sports  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


  6%|▌         | 3/50 [00:09<02:22,  3.02s/it]

Index 2: Fail 🧐 
dataset index: 1829
Confidence: 22.19
NLL: 82.29
Failed Query: magic number down to 10 New York -- If the Braves #39; 13th consecutive division title seemed like a foregone conclusion before Wednesday, well then it seems doubly so today.
Label :  Sports  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


  8%|▊         | 4/50 [00:12<02:12,  2.88s/it]

Index 3: Fail 🧐 
dataset index: 3906
Confidence: 27.74
NLL: 63.88
Failed Query: 14 dead in Kashmir violence NEW DELHI - Fourteen people have been killed in Kashmir in an increase of violence since a visit by the Indian prime minister in mid-November.
Label :  World  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 10%|█         | 5/50 [00:26<05:11,  6.92s/it]

Index 4: Fail 🧐 
dataset index: 5307
Confidence: 28.31
NLL: inf
Failed Query: Microsoft Plans Heavy Hype for 'Halo 2' "Halo 2" appears to be one of the most hotly hyped and heavily anticipated video games ever, and Microsoft is planning a Tuesday release that may rival the best of Hollywood's movie glitz. &lt;FONT face="verdana,MS Sans Serif,arial,helvetica" size="-2"\ color="#666666"&gt;&lt;B&gt;-washingtonpost.com&lt;/B&gt;&lt;/FONT&gt;
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 12%|█▏        | 6/50 [00:28<03:59,  5.44s/it]

Index 5: Fail 🧐 
dataset index: 3621
Confidence: 16.62
NLL: 78.16
Failed Query: Adams quits as Leicester boss LEICESTER: Mickey Adams has quit as manager of English Championship side Leicester City, the club announced yesterday. Adams #39; resignation was accepted at an emergency meeting of the board of directors at the 
Label :  Sports  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 14%|█▍        | 7/50 [00:31<03:19,  4.63s/it]

Index 6: Fail 🧐 
dataset index: 3865
Confidence: 23.71
NLL: 83.70
Failed Query: WEEK 6: Pregame show hits the road for Seattle-New England Every week, the experts of FOX NFL Sunday will candidly reveal their observations and make their opinions known as they prepare for their top-rated pregame telecast - seen each Sunday at 12 pm ET / 9 am PT.
Label :  Sports  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 16%|█▌        | 8/50 [00:34<02:48,  4.02s/it]

Index 7: Fail 🧐 
dataset index: 3301
Confidence: 27.58
NLL: 63.52
Failed Query: Guinea-Bissau Soldiers Stage Mutiny, Kill Army Chief Mutinous soldiers demanding pay for peacekeeping duty abroad killed the commander of Guinea-Bissau #39;s armed forces on Wednesday and seized key buildings in the capital of the former Portuguese colony.
Label :  World  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 18%|█▊        | 9/50 [00:36<02:17,  3.36s/it]

Index 8: Fail 🧐 
dataset index: 5565
Confidence: 29.98
NLL: inf
Failed Query:  #39;Star Trek #39; spaceship enters the gateway to the Moon A EUROPEAN spacecraft powered by a Star Trek-style thruster has flown through a lunar gateway that puts it on course to reach the Moon on Monday.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 0.00%
----------------------------------


 20%|██        | 10/50 [00:39<02:07,  3.18s/it]

Index 9: Success 🎉  
dataset index: 7488
Confidence: 31.19
NLL: -0.00
Current Accuracy: 10.00%
----------------------------------


 22%|██▏       | 11/50 [00:43<02:20,  3.61s/it]

Index 10: Fail 🧐 
dataset index: 4569
Confidence: 28.06
NLL: inf
Failed Query: Aids warning over bushmeat trade Meat from African wild animals being sold illegally in the UK is spreading a virus similar to HIV, a leading scientist warns.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 9.09%
----------------------------------


 24%|██▍       | 12/50 [00:46<02:06,  3.32s/it]

Index 11: Fail 🧐 
dataset index: 2871
Confidence: 16.21
NLL: 74.65
Failed Query: Born to coach For the reason, with apologies to Michael Vick, look no further than the third-youngest head coach in the NFL. James Lawrence Mora, the son, is already starting to look suspiciously like father James Earnest 
Label :  Sports  Pred :  Business 
Current Accuracy: 8.33%
----------------------------------


 26%|██▌       | 13/50 [00:48<01:50,  2.98s/it]

Index 12: Fail 🧐 
dataset index: 7475
Confidence: 26.83
NLL: 84.34
Failed Query: Efficient Germany Sweep Past Japan Two goals from Miroslav Klose helped Jrgen Klinsmann #39;s experimental Germany side breeze to a 3-0 win over Japan in Yokohama this afternoon.
Label :  Sports  Pred :  Business 
Current Accuracy: 7.69%
----------------------------------


 28%|██▊       | 14/50 [00:51<01:42,  2.85s/it]

Index 13: Fail 🧐 
dataset index: 143
Confidence: 27.53
NLL: 63.38
Failed Query: Israelis OK new homes in West Bank A leaked Israeli plan to build 1,000 new Jewish settler homes in the West Bank yesterday sent Bush administration officials scrambling for a response in the sensitive period before November #39;s presidential election. 
Label :  World  Pred :  Business 
Current Accuracy: 7.14%
----------------------------------


 30%|███       | 15/50 [00:53<01:36,  2.77s/it]

Index 14: Fail 🧐 
dataset index: 5954
Confidence: 29.69
NLL: 91.84
Failed Query: Delhomme Questionable Against Cardinals (AP) AP - Carolina quarterback Jake Delhomme did not practice Wednesday and is questionable this week because of a broken bone in his right thumb.
Label :  Sports  Pred :  Business 
Current Accuracy: 6.67%
----------------------------------


 32%|███▏      | 16/50 [00:56<01:36,  2.83s/it]

Index 15: Fail 🧐 
dataset index: 6768
Confidence: 27.28
NLL: inf
Failed Query: Japan #39;s NTT DoCoMo sees Europe embracing hi-tech mobile phones TOKYO : Japan #39;s top mobile operator NTT DoCoMo believes Europe will embrace hi-tech telephones and expects a major boost in subscribers on the continent of its i-mode Internet service.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 6.25%
----------------------------------


 34%|███▍      | 17/50 [00:59<01:32,  2.80s/it]

Index 16: Fail 🧐 
dataset index: 4298
Confidence: 28.59
NLL: 88.29
Failed Query: VIEIRA FIT FOR UNITED BATTLE SKIPPER Patrick Vieira is set to hand Arsenal a massive boost for Sunday #39;s crunch clash against Manchester United by declaring himself fit to lead the Gunners at Old Trafford.
Label :  Sports  Pred :  Business 
Current Accuracy: 5.88%
----------------------------------


 36%|███▌      | 18/50 [01:02<01:31,  2.86s/it]

Index 17: Success 🎉  
dataset index: 6222
Confidence: 26.81
NLL: -0.00
Current Accuracy: 11.11%
----------------------------------


 38%|███▊      | 19/50 [01:04<01:22,  2.68s/it]

Index 18: Success 🎉  
dataset index: 1705
Confidence: 25.97
NLL: -0.00
Current Accuracy: 15.79%
----------------------------------


 40%|████      | 20/50 [01:07<01:18,  2.62s/it]

Index 19: Fail 🧐 
dataset index: 2158
Confidence: 27.28
NLL: 62.81
Failed Query: US Hostage Apparently Beheaded (CBS/AP) A video posted on an Islamic Web site Monday shows the apparent beheading of a man identified in the tape as American construction contractor Eugene Armstrong.
Label :  World  Pred :  Business 
Current Accuracy: 15.00%
----------------------------------


 42%|████▏     | 21/50 [01:09<01:13,  2.53s/it]

Index 20: Fail 🧐 
dataset index: 7502
Confidence: 25.47
NLL: 58.64
Failed Query: Sharon Predicts 'Breakthrough' in Ties With Palestinians Prime Minister Ariel Sharon said that his government would implement his proposal to dismantle all the Israeli settlements in Gaza and four small ones in the West Bank on schedule.
Label :  World  Pred :  Business 
Current Accuracy: 14.29%
----------------------------------


 44%|████▍     | 22/50 [01:12<01:13,  2.62s/it]

Index 21: Fail 🧐 
dataset index: 6123
Confidence: 21.49
NLL: 49.48
Failed Query: Van Gogh #39;s murder brings out Holland #39;s contradictions The murder of Dutch filmmaker Theo van Gogh by a young Muslim of Moroccan descent has shaken Holland to its very foundations. To most people, including the Dutch, the killing and its violent 
Label :  World  Pred :  Business 
Current Accuracy: 13.64%
----------------------------------


 46%|████▌     | 23/50 [01:14<01:08,  2.54s/it]

Index 22: Fail 🧐 
dataset index: 134
Confidence: 25.97
NLL: 88.22
Failed Query: Former Florida Swimming Coach Dies at 83 (AP) AP - William H. Harlan, the retired University of Florida swimming coach who led the Gators to eight conference titles, died Tuesday, school officials said. He was 83.
Label :  Sports  Pred :  Business 
Current Accuracy: 13.04%
----------------------------------


 48%|████▊     | 24/50 [01:17<01:06,  2.54s/it]

Index 23: Fail 🧐 
dataset index: 5884
Confidence: 25.48
NLL: 58.67
Failed Query: Argentina gets China investment China is to invest nearly \$20bn (11bn) in Argentina over the next 10 years, Argentine officials say.
Label :  World  Pred :  Business 
Current Accuracy: 12.50%
----------------------------------


 50%|█████     | 25/50 [01:20<01:05,  2.62s/it]

Index 24: Fail 🧐 
dataset index: 837
Confidence: 23.68
NLL: 77.52
Failed Query: Helm #39;s perfect 10 And the two-and-a-half back somersaults with one and a half twists in a pike position turned out to be his ticket to a silver medal.
Label :  Sports  Pred :  Business 
Current Accuracy: 12.00%
----------------------------------


 52%|█████▏    | 26/50 [01:22<01:01,  2.58s/it]

Index 25: Fail 🧐 
dataset index: 6688
Confidence: 28.78
NLL: 88.79
Failed Query: Browns coach Davis resigns; Robiskie named interim coach BEREA, Ohio (Ticker) - One day after admitting that his shaky job status was a  quot;distraction quot; to his players, Butch Davis resigned as coach of the Cleveland Browns on Tuesday.
Label :  Sports  Pred :  Business 
Current Accuracy: 11.54%
----------------------------------


 54%|█████▍    | 27/50 [01:25<00:59,  2.59s/it]

Index 26: Fail 🧐 
dataset index: 6722
Confidence: 29.10
NLL: 87.42
Failed Query: India #39;s Nehra to miss Bangladesh series Left-arm seamer Ashish Nehra has been left out of India #39;s squad for this month #39;s two-test series against Bangladesh due to an abdominal strain.
Label :  Sports  Pred :  Business 
Current Accuracy: 11.11%
----------------------------------


 56%|█████▌    | 28/50 [01:27<00:54,  2.48s/it]

Index 27: Success 🎉  
dataset index: 3861
Confidence: 27.22
NLL: -0.00
Current Accuracy: 14.29%
----------------------------------


 58%|█████▊    | 29/50 [01:29<00:47,  2.28s/it]

Index 28: Success 🎉  
dataset index: 6997
Confidence: 29.47
NLL: -0.00
Current Accuracy: 17.24%
----------------------------------


 60%|██████    | 30/50 [01:31<00:46,  2.33s/it]

Index 29: Fail 🧐 
dataset index: 1521
Confidence: 23.04
NLL: 53.05
Failed Query: Alleged U.S. Deserter Set to Surrender TOKYO - Accused U.S. Army deserter Charles Jenkins left his Tokyo hospital for an American military base to surrender to military authorities Saturday, nearly 40 years after he allegedly defected to North Korea...
Label :  World  Pred :  Business 
Current Accuracy: 16.67%
----------------------------------


 62%|██████▏   | 31/50 [01:34<00:45,  2.38s/it]

Index 30: Success 🎉  
dataset index: 2988
Confidence: 25.21
NLL: -0.00
Current Accuracy: 19.35%
----------------------------------


 64%|██████▍   | 32/50 [01:36<00:39,  2.21s/it]

Index 31: Fail 🧐 
dataset index: 3261
Confidence: 31.06
NLL: 90.33
Failed Query: Hewitt Advances at Japan Open (AP) AP - Top seeded Lleyton Hewitt rallied to a 6-0, 3-6, 6-1 win over Japan's Gouichi Motomura on Wednesday in the second round of the Japan Open.
Label :  Sports  Pred :  Business 
Current Accuracy: 18.75%
----------------------------------


 66%|██████▌   | 33/50 [01:39<00:42,  2.47s/it]

Index 32: Fail 🧐 
dataset index: 1756
Confidence: 30.39
NLL: 69.98
Failed Query: Report: Japan PM Advisers See China Military Threat (Reuters) Reuters - In a move that could further chill ties\between the two Asian powers, an advisory panel to Japan's\prime minister will say China should be described as a military\threat in a defense review, the Nihon Keizai newspaper reported\on Wednesday.
Label :  World  Pred :  Business 
Current Accuracy: 18.18%
----------------------------------


 68%|██████▊   | 34/50 [01:41<00:40,  2.51s/it]

Index 33: Success 🎉  
dataset index: 2816
Confidence: 10.92
NLL: -0.00
Current Accuracy: 20.59%
----------------------------------


 70%|███████   | 35/50 [01:44<00:37,  2.50s/it]

Index 34: Fail 🧐 
dataset index: 7041
Confidence: 24.51
NLL: 56.44
Failed Query:  #39;Miracle in mud #39; as four pulled alive from Philippine disaster REAL, Philippines (AFP) - Philippine rescuers were frantically digging for more survivors after four people, including a toddler, were pulled alive from a building crushed by a landslide 11 days ago.
Label :  World  Pred :  Business 
Current Accuracy: 20.00%
----------------------------------


 72%|███████▏  | 36/50 [01:46<00:34,  2.48s/it]

Index 35: Fail 🧐 
dataset index: 3365
Confidence: 27.30
NLL: 84.25
Failed Query: Game On On the same day that Tiger Woods was married in Barbados, the two best golfers in the world were walking the streets of St. Andrews, Scotland.
Label :  Sports  Pred :  Business 
Current Accuracy: 19.44%
----------------------------------


 74%|███████▍  | 37/50 [01:49<00:33,  2.55s/it]

Index 36: Fail 🧐 
dataset index: 6660
Confidence: 30.43
NLL: 95.37
Failed Query: Prosecutor: Pacers Players to Be Charged (AP) AP - Indiana Pacers players will be charged for fighting with fans during the Nov. 19 brawl at the end of a game against the Detroit Pistons, Oakland County Prosecutor David Gorcyca told The Detroit News.
Label :  Sports  Pred :  Business 
Current Accuracy: 18.92%
----------------------------------


 76%|███████▌  | 38/50 [01:56<00:48,  4.07s/it]

Index 37: Fail 🧐 
dataset index: 4011
Confidence: 31.31
NLL: 93.34
Failed Query: Manning Throws 3 TDs to Lead Colts Past Titans  INDIANAPOLIS (Sports Network) - Peyton Manning threw for  425 yards with three touchdowns and Edgerrin James ran for 105  yards with a pair of scores, as the Indianapolis Colts shook  off a sluggish start and rolled to a 51-24 victory over the  Tennessee Titans at the RCA Dome.
Label :  Sports  Pred :  Business 
Current Accuracy: 18.42%
----------------------------------


 78%|███████▊  | 39/50 [01:59<00:39,  3.57s/it]

Index 38: Fail 🧐 
dataset index: 4945
Confidence: 24.90
NLL: 57.32
Failed Query: Profile: Leftist Vazquez wins Uruguay presidential poll Uruguay #39;s leftist candidate Tabare Vazquez won a historic victory in Sunday #39;s presidential elections with more than 50 percent of the ballot.
Label :  World  Pred :  Business 
Current Accuracy: 17.95%
----------------------------------


 80%|████████  | 40/50 [02:01<00:31,  3.14s/it]

Index 39: Success 🎉  
dataset index: 7556
Confidence: 28.91
NLL: -0.00
Current Accuracy: 20.00%
----------------------------------


 82%|████████▏ | 41/50 [02:03<00:26,  2.95s/it]

Index 40: Fail 🧐 
dataset index: 1609
Confidence: 28.95
NLL: inf
Failed Query: Bullies move online Singapore - More complaints of cyberbullying are emerging from youngsters in Singapore than any other country except the United States, an international safety group said in a report on Monday.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 19.51%
----------------------------------


 84%|████████▍ | 42/50 [02:06<00:22,  2.82s/it]

Index 41: Success 🎉  
dataset index: 405
Confidence: 28.44
NLL: -0.00
Current Accuracy: 21.43%
----------------------------------


 86%|████████▌ | 43/50 [02:09<00:19,  2.77s/it]

Index 42: Fail 🧐 
dataset index: 1908
Confidence: 30.46
NLL: 87.55
Failed Query: Kluivert gives Souness stuttering start at Newcastle Patrick Kluivert struck twice as Graeme Souness began his reign at St James Park with a 2-0 win over Israeli Arab side Bnei Sakhnin in the UEFA Cup first round, first leg at Newcastle United this morning.
Label :  Sports  Pred :  Business 
Current Accuracy: 20.93%
----------------------------------


 88%|████████▊ | 44/50 [02:11<00:15,  2.64s/it]

Index 43: Fail 🧐 
dataset index: 2222
Confidence: 26.56
NLL: inf
Failed Query: Wireless carriers: Privacy bill not needed WASHINGTON - Representatives of wireless telephone carriers planning a telephone directory service told a U.S. Senate committee Tuesday that legislation to protect their customers' privacy isn't needed, because their plan already does.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 20.45%
----------------------------------


 90%|█████████ | 45/50 [02:14<00:13,  2.68s/it]

Index 44: Success 🎉  
dataset index: 6878
Confidence: 18.36
NLL: -0.00
Current Accuracy: 22.22%
----------------------------------


 92%|█████████▏| 46/50 [02:16<00:10,  2.67s/it]

Index 45: Fail 🧐 
dataset index: 3781
Confidence: 31.88
NLL: 87.36
Failed Query: Ruud double lifts Dutch Amsterdam - Goals in either half from Manchester United striker Ruud van Nistelrooy lifted the Netherlands to a 3-1 win over Finland in their European Group One qualifier for the 2006 World Cup here on Wednesday.
Label :  Sports  Pred :  Business 
Current Accuracy: 21.74%
----------------------------------


 94%|█████████▍| 47/50 [02:19<00:07,  2.52s/it]

Index 46: Fail 🧐 
dataset index: 4643
Confidence: 27.28
NLL: inf
Failed Query: NASA expert says Bush stifles evidence on global warming IOWA CITY, Iowa A NASA scientist has charged that the Bush administration is subverting science and misleading the public by trying to suppress or alter evidence on the dangers of global warming.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 21.28%
----------------------------------


 96%|█████████▌| 48/50 [02:21<00:05,  2.62s/it]

Index 47: Fail 🧐 
dataset index: 5190
Confidence: 23.45
NLL: 100.98
Failed Query: Research predicts iceless Arctic Global warming is causing the Arctic ice-cap to melt at such an unprecedented rate that by the summer of 2070 it may have no ice at all, according to the most comprehensive study carried out on global climate change in the region.
Label :  Sci/Tech  Pred :  Business 
Current Accuracy: 20.83%
----------------------------------


 98%|█████████▊| 49/50 [02:29<00:04,  4.02s/it]

Index 48: Fail 🧐 
dataset index: 5941
Confidence: 27.72
NLL: 63.83
Failed Query: Cricket: NZ suffer Franklin blow New Zealand bowler James Franklin misses the first Test against Australia with  injury.
Label :  World  Pred :  Business 
Current Accuracy: 20.41%
----------------------------------


100%|██████████| 50/50 [02:32<00:00,  3.04s/it]

Index 49: Fail 🧐 
dataset index: 580
Confidence: 29.07
NLL: 94.36
Failed Query: Bryant Prosecutors Question Defense DNA Evidence  DENVER (Reuters) - Prosecutors in the rape case against  U.S. basketball star Kobe Bryant are questioning the validity  of DNA evidence crucial to the defense's case, saying data  appeared to have been manipulated and might have to be thrown  out.
Label :  Sports  Pred :  Business 
Current Accuracy: 20.00%
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Total Correct Predictions: 10 (20.00%)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>


---


# 실습: Zero-shot Classification

이번 실습에서는 open LLM을 가지고 zero-shot classification을 해봅니다. 먼저 필요한 library들을 설치합시다.


그 다음 Gemma-2B를 사용하기 위해 다음과 같은 작업을 진행합니다:

1. huggingface.co 계정 만들고 로그인하기
2. https://www.kaggle.com/models/google/gemma/license/consent 에서 Gemma license 동의하기
3. 홈 화면으로 돌아와, `Profile > Settings > Access Tokens` 메뉴로 들어와 "Write" type의 token 생성하기
4. 생성한 토큰을 아래 "HF TOKEN"에 불여넣고 셀을 실행하기.


In [ ]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")

이번에는 Gemma-2B를 가지고 간단한 text 생성을 해봅시다.
"What is your name?" 이라는 text를 넣었을 때 어떤 text가 생성되는지 살펴봅시다.


In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
input_ids

In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
outputs = model.generate(**input_ids)
print(tokenizer.decode(outputs[0]))

In [ ]:
tokens = input_ids["input_ids"]
print(tokens)
input_ids

2B의 작은 LLM이라 질좋은 답변이 나오지 않는 것을 알 수 있습니다.
이번에는 입력으로 넣어준 token들의 logit을 계산해봅시다.


In [ ]:
tokens = input_ids["input_ids"]
print(tokens)

logits = model(**input_ids).logits
for i in range(tokens.shape[-1]):
    token = tokens[0, i].item()
    print(logits[0, i, token])

위와 같이 모델 출력의 `.logits`을 통해 token들의 logit을 알 수 있습니다.
Logit은 높을 수록 token이 나올 확률이 높다는 뜻입니다.

이번에는 logit 계산을 통해 zero-shot classification을 구현해보도록 하겠습니다.


In [8]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.cuda.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

아래는 실제로 zero-shot classification을 해본 결과입니다.


In [ ]:
probs = zero_shot_classification(
    "I am happy!", "Is the sentence positive or negative?: ", ["positive", "negative"]
)
print(probs)

보시다시피 우리는 Gemma를 별도로 학습하지 않았음에도 불구하고 주어진 문장이 긍정적이라는 것을 정확하게 예측하고 있습니다.

다음은 영화 리뷰 감정 분석 task에 적용해봅시다.
먼저 data를 불러옵니다.


In [ ]:
from datasets import load_dataset


imdb = load_dataset("imdb")


def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_imdb = imdb.map(preprocess_function, batched=True)

그리고 `test` data에서 50개의 영화 리뷰에 대해 예측하는 코드는 다음과 같습니다.


In [ ]:
import numpy as np
from tqdm import tqdm


n_corrects = 0
for i in tqdm(range(50)):
    text = tokenized_imdb["test"][i]["text"]
    label = tokenized_imdb["test"][i]["label"]
    probs = zero_shot_classification(
        text,
        "A movie review is given. Decide that the movie review is positive or negative: ",
        labels=["Answer: negative.", "Answer: positive."],
    )

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1

print(n_corrects)

In [ ]:
text = tokenized_imdb["test"][i]["text"]
text

In [ ]:
label = tokenized_imdb["test"][i]["label"]
label

보시다시피 정확도 88%로, 매우 높은 성능을 보이는 것을 알 수 있습니다.
